src/execution/compiler/operator/hash_aggregation_translator.cpp

```diff
@@ -250,24 +263,131 @@ void HashAggregationTranslator::TearDownAggregationHashTable(FunctionBuilder *fu
+  if (IsBuildPipeline(pipeline)) {
+    if (build_pipeline_.IsParallel()) {
+      TearDownAggregationHashTable(function, local_agg_ht_.GetPtr(GetCodeGen()));
+    }
+  }
+}
+
+void HashAggregationTranslator::RecordCounters(const Pipeline &pipeline, FunctionBuilder *function) const {
+  auto *codegen = GetCodeGen();
+  if (IsBuildPipeline(pipeline)) {
+    const auto &agg_ht = pipeline.IsParallel() ? local_agg_ht_ : global_agg_ht_;
+    FeatureRecord(function, brain::ExecutionOperatingUnitType::AGGREGATE_BUILD,
+                  brain::ExecutionOperatingUnitFeatureAttribute::NUM_ROWS, pipeline, CounterVal(num_agg_inputs_));
+
+    if (build_pipeline_.IsParallel() && IsCountersEnabled()) {
+      // In parallel mode, we subtract from the insert count of the last task that might
+      // have been run with the current thread. This is to more correctly model the
+      // amount of work performed by the current task.
+      auto *ins_count = codegen->CallBuiltin(ast::Builtin::AggHashTableGetInsertCount, {agg_ht.GetPtr(codegen)});
+      auto *minus = codegen->BinaryOp(parsing::Token::Type::MINUS, ins_count, agg_count_.Get(codegen));
+      FeatureRecord(function, brain::ExecutionOperatingUnitType::AGGREGATE_BUILD,
+                    brain::ExecutionOperatingUnitFeatureAttribute::CARDINALITY, pipeline, minus);
+    } else {
+      FeatureRecord(function, brain::ExecutionOperatingUnitType::AGGREGATE_BUILD,
+                    brain::ExecutionOperatingUnitFeatureAttribute::CARDINALITY, pipeline,
+                    codegen->CallBuiltin(ast::Builtin::AggHashTableGetTupleCount, {agg_ht.GetPtr(codegen)}));
+    }
+
+    FeatureArithmeticRecordMul(function, pipeline, GetTranslatorId(), CounterVal(num_agg_inputs_));
+  } else {
+    ast::Expr *agg_ht;
+    if (pipeline.IsParallel()) {
+      agg_ht = iterate_agg_ht_.Get(codegen);
+    } else {
+      agg_ht = global_agg_ht_.GetPtr(codegen);
+    }
+
+    FeatureRecord(function, brain::ExecutionOperatingUnitType::AGGREGATE_ITERATE,
+                  brain::ExecutionOperatingUnitFeatureAttribute::NUM_ROWS, pipeline, CounterVal(num_agg_outputs_));
+
+    FeatureRecord(function, brain::ExecutionOperatingUnitType::AGGREGATE_ITERATE,
+                  brain::ExecutionOperatingUnitFeatureAttribute::CARDINALITY, pipeline,
+                  codegen->CallBuiltin(ast::Builtin::AggHashTableGetTupleCount, {agg_ht}));
+    FeatureArithmeticRecordMul(function, pipeline, GetTranslatorId(), CounterVal(num_agg_outputs_));
+  }
+}
+
+void HashAggregationTranslator::EndParallelPipelineWork(const Pipeline &pipeline, FunctionBuilder *function) const {
+  RecordCounters(pipeline, function);
+
+  if (IsBuildPipeline(pipeline) && IsCountersEnabled()) {
+    auto *codegen = GetCodeGen();
+    const auto &agg_ht = local_agg_ht_;
+    auto *tuple_count = codegen->CallBuiltin(ast::Builtin::AggHashTableGetInsertCount, {agg_ht.GetPtr(codegen)});
+    function->Append(codegen->Assign(agg_count_.Get(codegen), tuple_count));
```

Comment:
That is accurate. This is mainly done since TBB can assign multiple tasks using the same thread-local state so it is very likely to change if and when TBB gets replaced. This value is still a little bit dubious in general --- since it tracks the number of unique inserts from the last "state". That is also why `agg_count_` cannot be set to 0 in `InitializeCounters` (since there is also an implicit guarantee that `BeginParallelPipelineWork` will call `InitializeCounters` which might also benefit from more visible documentation).

I think I will add documentation around the `agg_count_` variable in `src/include/execution/compiler/operator/hash_aggregation_translator.h` (along with references) unless you have a better place in mind.

Reply:
Docs have been added.